session/session.go

```diff
@@ -1203,6 +1203,11 @@ func runStmt(ctx context.Context, se *session, s sqlexec.Statement) (rs sqlexec.
 	rs, err = s.Exec(ctx)
 	sessVars.TxnCtx.StatementCount++
 	if !s.IsReadOnly(sessVars) {
+		// Handle the stmt panic in a transaction.
+		if err != nil {
+			se.txn.doNotCommit = err
+		}
+
```

Comment:
@bobotu @XuHuaiyu 
I've tested it in transaction or non-transaction, it works.
```
// with transaction
drop table if exists t;
create table t(a int);
begin;
insert into t values(1), (-1), (0);
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
tidb> insert /*+ memory_quota(1 MB) */ into t select * from t;
ERROR 1105 (HY000): Out Of Memory Quota![conn_id=1]
tidb> insert /*+ memory_quota(2 MB) */ into t select * from t;
ERROR 1105 (HY000): current transaction is aborted, commands ignored until end of transaction block:Out Of Memory Quota![conn_id=1]
tidb> insert /*+ memory_quota(14 MB) */ into t select * from t;
ERROR 1105 (HY000): current transaction is aborted, commands ignored until end of transaction block:Out Of Memory Quota![conn_id=1]
tidb> insert /*+ memory_quota(15 MB) */ into t select * from t;
ERROR 1105 (HY000): current transaction is aborted, commands ignored until end of transaction block:Out Of Memory Quota![conn_id=1]
tidb> commit;
ERROR 1105 (HY000): Out Of Memory Quota![conn_id=1]
tidb> select count(*) from t;
+----------+
| count(*) |
+----------+
|        0 |
+----------+
1 row in set (0.00 sec)


// without transaction
drop table if exists t;
create table t(a int);
insert into t values(1), (-1), (0);
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
tidb> insert /*+ memory_quota(1 MB) */ into t select * from t;
ERROR 1105 (HY000): Out Of Memory Quota![conn_id=1]
tidb> insert /*+ memory_quota(2 MB) */ into t select * from t;
ERROR 1105 (HY000): Out Of Memory Quota![conn_id=1]
tidb> insert /*+ memory_quota(14 MB) */ into t select * from t;
Query OK, 393216 rows affected (1.62 sec)
Records: 393216  Duplicates: 0  Warnings: 0
tidb> insert /*+ memory_quota(15 MB) */ into t select * from t;
ERROR 1105 (HY000): Out Of Memory Quota![conn_id=1]
tidb> select count(*) from t;
+----------+
| count(*) |
+----------+
|   786432 |
+----------+
1 row in set (0.09 sec)
```

Reply:
You may follow @bobotu 's suggestion that recovers in `stmtCommit`

Reply:
This change is incorrect. This will invalidate all transactions even it is valid. The correct way to fix the original issue is add `recover` in `StmtCommit`.